File: src/Heuristic.cpp
#include <Heuristic.hpp>

using namespace std;

Heuristic::Heuristic(Data &data, int nb_threads, int time_limit_complete, int time_limit_per_combination)
{
    this->time_limit_complete = time_limit_complete;
    this->time_limit_per_combination = time_limit_per_combination;

    this->nb_threads = nb_threads;

    start = chrono::steady_clock::now();
    execute_heuristic(data);
    end = chrono::steady_clock::now();

    cout << endl << "-> Total time = " << chrono::duration<double>(end-start).count() << endl;
    if (!overall_best_sol.feasible)
    {
        cout << "No feasible solution was found with heuristic method!" << endl;
    }
    else
    {
        // display best solution
        overall_best_sol.display_solution(data);
    }
}

void Heuristic::execute_heuristic(Data &data)
{
    create_initial_candidates(data);
    cout << "Solving initial candidates..." << endl;
    solve_initial_candidates(data);
    while (improved_sol)
    {
        
        iter++;

        // verify time limit
        end = chrono::steady_clock::now();
        chrono::duration<double> current_time = end-start;
        if (current_time.count() >= time_limit_complete)
        { 
            reached_time_limit(data, current_time);
        }

        // create subsets of candidate combinations
        bool valid_subsets = create_subsets_from_best_combination(data);
        if (!valid_subsets)
            break;

        cout << "Updating candidates - iter = " << iter << endl;

        // update candidates with subsets
        execute_candidate_combinations(data);
    }
    cout << "Terminating loop at iteration " << iter << endl;

    // unfix final trips
    unfix_trips_from_best_combination(data);
    ModelORTools model;
    model.initialize(data, 1);
    model.create_model_with_routes_constraints(data, overall_best_sol.routes_combination);
    model.execute_solver_for_combination(data, overall_best_sol.obj_value, time_limit_per_combination, "heuristic");
}

void Heuristic::execute_candidate_combinations(Data &data)
{
    // vectors with model objects for each thread
    vector <ModelORTools> models(nb_threads);

    improved_sol = false;
    int counter_solved = 0;

    #pragma omp parallel num_threads(nb_threads)
    {
        int thread_id = omp_get_thread_num();
        #pragma omp for schedule(dynamic)
        for (unsigned long long count = 0; count < candidate_combinations.size(); count++)
        {
            // get current combination
            RoutesCombination current_combination = candidate_combinations[count];

            if (comb.normalize_combination(data, current_combination))
            {
                ModelORTools &model_thread = models[thread_id];
                model_thread.initialize(data, 1);
                model_thread.create_model_with_routes_constraints(data, current_combination);
                int feasible = model_thread.execute_solver_for_combination(data, overall_best_sol.obj_value, time_limit_per_combination, "heuristic");
                if (feasible)
                {
                    #pragma omp critical
                    {
                        if (model_thread.best_sol.obj_value < overall_best_sol.obj_value)
                        {
                            overall_best_sol = model_thread.best_sol;
                            improved_sol = true;
                        }
                        else if (model_thread.best_sol.obj_value == overall_best_sol.obj_value)
                        {
                            if (model_thread.best_sol.max_nb_repeated_routes > overall_best_sol.max_nb_repeated_routes)
                            {
                                overall_best_sol = model_thread.best_sol;
                                improved_sol = true;
                            }
                        }
                    }
                }
            }
            #pragma omp critical
            {
                counter_solved++;
                cout << counter_solved << "/" << candidate_combinations.size() << " candidate combination(s) tested! (Thread " << thread_id << ")" << endl;
            }

            // verify time limit
            chrono::time_point<std::chrono::steady_clock> now = chrono::steady_clock::now();
            chrono::duration<double> current_time = now-start;
            if (current_time.count() >= time_limit_complete)
            {
                #pragma omp critical
                {
                    reached_time_limit(data, current_time);
                }
            }
        }
    }
}

void Heuristic::solve_initial_candidates(Data &data)
{
    execute_candidate_combinations(data);
    if (improved_sol == false)
    {
        // create subsets from all candidates combinations
        if (create_subsets_from_all_candidates(data))
        {
            execute_candidate_combinations(data);
            if (improved_sol == false)
            {
                // update set of valid routes
                try_new_set_of_routes(data);
                execute_candidate_combinations(data);
                if (improved_sol == false)
                {
                    // allow model to choose the last trips completed by the trains
                    unfix_trips_from_best_combination(data);
                    execute_candidate_combinations(data);
                }
            }
        }
    }
}

void Heuristic::reached_time_limit(Data &data, std::chrono::duration<double> time)
{
    cout << endl << ">> Reached time limit!" << endl;
    cout << "\tTerminating process!" << endl;
    if (overall_best_sol.feasible == false)
    {
        cout << "\tNo feasible solution was found..." << endl;
    }
    cout << endl << "-> Total time = " << time.count() << endl;
    exit(0);
}

void Heuristic::create_initial_candidates (Data &data)
{
    create_cyclical_routes_set(data);
    create_initial_valid_routes_set(data);

    cout << endl << endl << "Creating initial set of candidate combinations..." << endl;

    // create maximum size combinations
    create_maximum_size_candidates(data);
}

void Heuristic::create_maximum_size_candidates(Data &data)
{
    int nb_trains = data.get_nb_trains();
    int num_valid = initial_valid_routes_set.size();
    int num_cyclic = cyclical_routes_set.size();

    // map initial routes done by each train
    vector<int> initial_indices(nb_trains, 0);
    bool done_initial = false;
    while (!done_initial)
    {
        // map cyclical routes done by each train
        vector<int> cyclic_indices(nb_trains, 0);
        bool done_cyclic = false;
        while (!done_cyclic)
        {
            vector<vector<int>> combination;
            bool valid_combination = true;
            for (int t = 0; t < nb_trains; t++)
            {
                int nb_trips = data.get_train_max_trips(t);
                vector<int> routes_of_train(nb_trips);

                // choose initial trip
                routes_of_train[0] = initial_valid_routes_set[initial_indices[t]];

                // choose cyclical route
                for (int k = 1; k < nb_trips; k++)
                {
                    routes_of_train[k] = cyclical_routes_set[cyclic_indices[t]];
                }

                // verify whether sequence of routes is compatible
                if (verify_compatibility(data, routes_of_train))
                {
                    combination.push_back(routes_of_train);
                }
                else
                {
                    valid_combination = false;
                }
            }

            // verify whether demands are met
            // pending: rever essa parte
            if (valid_combination && verify_demands(data, combination) && normalize_candidates(data, combination))
            {
                candidate_combinations.push_back(combination);
            }

            if (data.get_max_nb_trips() == 1)
                break;

            // increment cyclical indeces
            for (int i = nb_trains - 1; i >= 0; i--)
            {
                if (++cyclic_indices[i] < num_cyclic)
                {
                    break;
                }
                else
                {
                    cyclic_indices[i] = 0;
                    if (i == 0) done_cyclic = true;
                }
            }
        }

        // increment initial indices
        for (int i = nb_trains - 1; i >= 0; i--)
        {
            if (++initial_indices[i] < num_valid)
            {
                break;
            }
            else
            {
                initial_indices[i] = 0;
                if (i == 0) done_initial = true;
            }
        }
    }
}

// functions to create subsets from routes combinations

bool Heuristic::create_subsets_from_best_combination(Data &data)
{
    vector<RoutesCombination> new_candidates;
    remove_trips(data, new_candidates, overall_best_sol.routes_combination, data.get_max_nb_trips()-iter);

    if (new_candidates.empty())
    {
        return false;
    }
    else
    {   candidate_combinations = new_candidates;
        return true;
    }
}

bool Heuristic::create_subsets_from_all_candidates(Data &data)
{
    vector<vector<vector<int>>> new_candidates;
    for (int i = 0; i < candidate_combinations.size(); i++)
    {
        remove_trips(data, new_candidates, candidate_combinations[i], data.get_max_nb_trips()-iter);
    }

    if (new_candidates.empty())
    {
        return false;
    }
    else
    {   candidate_combinations = new_candidates;
        return true;
    }
}

void Heuristic::remove_trips (Data &data, vector<RoutesCombination> &new_candidates, RoutesCombination &current_combination, int max_nb_trips)
{
    size_t total_combinations = 1 << data.get_nb_trains();     // 2^n possibilities
    for (size_t mask = 0; mask < total_combinations; mask++)
    {
        if (mask == 0) continue;
        RoutesCombination new_combination;

        bool valid = true;
        for (size_t i = 0; i < data.get_nb_trains(); i++)
        {
            // if bit is activated, try to remove it
            if ((mask >> i) & 1)
            {
                // verify if bit is from a trip that can be removed
                if (current_combination[i].size() < max_nb_trips)
                {
                    valid = false;
                    break;
                }
                new_combination.push_back(vector<int>(current_combination[i].begin(), current_combination[i].end() - 1));
            }
            else
            {
                new_combination.push_back(current_combination[i]);         // trips stay the same for the train
            }
        }

        // pending: modificar funcao que verifica se a combinação é válida e se atende as demandas
        if (valid && verify_demands(data, new_combination) && normalize_candidates(data, new_combination))
        {
            new_candidates.push_back(new_combination);
        }
    }
}

void Heuristic::unfix_trips_from_best_combination (Data &data)
{
    // last route is free for the model to choose
    for (int i = 0; i < data.get_nb_trains(); i++)
    {
        if (overall_best_sol.routes_combination[i].empty())
        {
            overall_best_sol.routes_combination[i].push_back(-1);
            continue;
        }
        overall_best_sol.routes_combination[i].back() = -1;
    }
}

void Heuristic::unfix_trips_from_all_candidates (Data &data)
{
    candidate_combinations.clear();
    int count = 0;
    for (const auto& tested : unique_combinations)
    {
        count++;
        RoutesCombination candidate = tested;
        // last route is free for the model to choose
        for (int i = 0; i < data.get_nb_trains(); i++)
        {
            if (candidate[i].empty())
            {
                candidate[i].push_back(-1);
                continue;
            }
            candidate[i].back() = -1;
        }
        if (normalize_candidates(data, candidate))
            candidate_combinations.push_back(candidate);
    }
}

void Heuristic::display_candidates_combinations()
{
    for (int i = 0; i < candidate_combinations.size(); i++)
    {
        cout << "Combination " << i+1 << ": " << endl;
        for (int j = 0; j < candidate_combinations[i].size(); j++)
        {
            cout << "Train " << j << ": ";
            for (int k = 0; k < candidate_combinations[i][j].size(); k++)
            {
                cout << candidate_combinations[i][j][k] << " ";
            }
            cout << endl;
        }
        cout << endl << endl;
    }
}

// ==================================================================== //

bool Heuristic::verify_compatibility (Data &data, vector<int> &current)
{
    bool flag = false; // flag to tell whether train completes any trips during the day
    for (int i = 0; i < current.size()-1; i++)
    {
        if (current[i] != data.get_nb_routes())
        {
            flag = true;
            if (current[i+1] != data.get_nb_routes())
            {
                // verify whether subsequential routes are compatible
                if (data.are_incompatible_routes(current[i], current[i+1])) return false;
            }
        }
        else if (current[i] == data.get_nb_routes())
        {
            // if trip is not made, verify whether the next ones also are not made
            if (i != current.size()-1 && current[i+1] != data.get_nb_routes()) return false;
        }
    }

    if (current.front() != data.get_nb_routes()) flag = true;

    return flag; 
}

bool Heuristic::normalize_candidates (Data &data, vector<vector<int>> &current)
{
    // normalize combinations to verify whether it was already tested previously
    // (only changes the train that will complete the trips)
    // return false if it was already tested and returns true if it's a new combination

    // normalize the vector
    vector<vector<int>> normalized_combination = current;
    sort(normalized_combination.begin(), normalized_combination.end());

    // verify whether it was already added according to the result
    pair<set<vector<vector<int>>>::iterator, bool> result;
    #pragma omp critical
    {
        result = unique_combinations.insert(normalized_combination);
    }
    if (!result.second)
    {
        return false; // combination already exists
    }
    return true; // current combination is a new combination
}

bool Heuristic::verify_demands(Data &data, vector<vector<int>> &current)
{
    // verify whether new subvector meets the demands
    int nb_trips_completed = 0;
    vector<int> times_vertex_was_visited (data.get_nb_vertices(), 0);
    for (int i = 0; i < current.size(); i++)
    {
        for (int j = 0; j < current[i].size(); j++)                        
        {
            nb_trips_completed++;
            int route = current[i][j];
            for (auto vertex : data.get_route_vertices(route))
            { 
                times_vertex_was_visited[vertex]++;
            }
        }
    }
    if (nb_trips_completed == 0) return false; 

    for (int i = 0 ; i < data.get_nb_vertices(); i++)
    {
        if (times_vertex_was_visited[i] < data.get_demand_per_day()[i]) return false;
    }
    return true;
}

void Heuristic::create_cyclical_routes_set(Data &data)
{
    cout << endl << "Selecting set of cyclical routes..." << endl;
    // // all cyclic routes included
    // for (int i = 0; i < data.get_nb_routes(); i++)
    // {
    //     if (data.is_cyclic_route(i))
    //         cyclical_routes_set.push_back(i);
    // }

    // search for full cycles included
    bool have_full_cycles = false;
    for (int i = 0; i < data.get_nb_routes(); i++)
    {
        if (data.get_route_vertices(i).size() == data.get_nb_vertices()+1)
        {
            cyclical_routes_set.push_back(i);
            have_full_cycles = true;
            have_cycles = true;
        }
    }

    if (!have_full_cycles)
    {
        for (int i = 0; i < data.get_nb_routes(); i++)
        {
            if ((data.is_cyclic_route(i)))
            {
                cyclical_routes_set.push_back(i);
                have_cycles = true;
            }
        }
    } 

    if (!have_cycles)
    {
        cout << "Instance does not have any cyclical routes..." << endl;
        exit(0);
    }
}

void Heuristic::create_initial_valid_routes_set(Data &data)
{
    cout << endl << "Selecting set of initial routes...";

    vector <int> will_be_removed (1, cyclical_routes_set.size()); // 0 for not being removed, 1 for being removed
    for (int i = 0; i < data.get_nb_routes(); i++)
    {
        if (data.is_valid_route(0, 0, i))
        {
            bool found_compatible = false;
            // verify whether initial route is compatible with any cyclical route before adding to set
            for (int j = 0; j < cyclical_routes_set.size(); j++)
            {
                if (!data.are_incompatible_routes(i, cyclical_routes_set[j]))
                {
                    found_compatible = true;
                    will_be_removed[j] = 0;
                }              
            }

            if (found_compatible) initial_valid_routes_set.push_back(i);
        }
    }

    // remove routes
    for (int i = 0; i < will_be_removed.size(); i ++)
    {
        if (will_be_removed[i] == 1)
        {
            cyclical_routes_set.erase(cyclical_routes_set.begin() + i);
        }
    }
}


void Heuristic::try_new_set_of_routes(Data &data)
{
    // add all cyclical routes
    cyclical_routes_set.clear();
    initial_valid_routes_set.clear();

    // all cyclic routes included
    for (int i = 0; i < data.get_nb_routes(); i++)
    {
        if (data.is_cyclic_route(i))
            cyclical_routes_set.push_back(i);
    }

    create_initial_valid_routes_set(data);
    create_maximum_size_candidates(data);
}

File: python/heuristic.py
from model import ModelTrainTimetabling
from solution import Solution
from combinations import Combinations

import time
import copy
import threading
from itertools import product, permutations, combinations_with_replacement
from concurrent.futures import ThreadPoolExecutor

class Heuristic:
    def __init__(self, data, nb_threads, time_limit_complete, time_limit_per_combination, solver):

        # initialize parameters
        self.data = data
        self.nb_threads = nb_threads
        self.time_limit_complete = time_limit_complete
        self.time_limit_per_combination = time_limit_per_combination
        self.solver = solver

        self.overall_best_sol = Solution()     # store the current best solution
        self.combinations = Combinations(data) # store the combinations to be explored

        self.full_cycle_routes = []            # store routes with full cycles (passes by all points of the railway)
        self.short_turn_routes = []            # store routes with short turns

        self.baseline_combinations = []        # baseline combinations, used to create reduced combinations with ledd trips
        self.candidate_combinations = []       # current candidate combinations

        self.route_min_time = None             # minimum duration of each route (used to calculate lower bound for combinations)

        self.time_limit_reached = False        # flag to tell if the time limit was reached
        self.start_time = None                 # store start time

        # variables used to administrate execution with multiple threads
        self.thread_local = threading.local()
        self.best_lock = threading.Lock()
        self.progress_lock = threading.Lock()
        self.time_limit_lock = threading.Lock()

# ===============================================================================================================
# =================================== functions used to calculate lower bound ===================================
# ===============================================================================================================        

    def compute_route_min_times(self):
        # minimum duration of each route: sum of minimum travel + service times of its arcs
        self.route_min_time = [sum(self.data.distance_and_service_min[arc["idx"]] for arc in self.data.route_arcs[r]) for r in range(self.data.nb_routes)]

    def combination_score(self, combination):
        # lower bound on the makespan: trains run in parallel, so the busiest train dominates
        train_times = [sum(self.route_min_time[r] for r in routes if 0 <= r < self.data.nb_routes) for routes in combination]
        return (max(train_times), sum(train_times))

# ===============================================================================================================
# ===============================================================================================================

# ===============================================================================================================
# ============================= functions used to create combinations according =================================
# ============================= to the number of trips                          =================================
# ===============================================================================================================

    def calculate_max_trips_per_train(self, S, max_trips):
        
        n = len(max_trips)
        result = []
        seen = set()

        def is_valid(comb):
            return all(comb[i] <= max_trips[i] for i in range(n))

        def main_comb(comb):
            # generate all valid permutations
            perms_valids = [
                p for p in set(permutations(comb))
                if is_valid(p)
            ]
            return min(perms_valids) # main combination

        def backtrack(pos, remaining, combination):
            if pos == n:
                if remaining == 0:
                    comb = main_comb(tuple(combination))
                    if comb not in seen:
                        seen.add(comb)
                        result.append(comb)
                return

            for i in range(0, max_trips[pos] + 1):
                if i > remaining:
                    break
                backtrack(pos + 1, remaining - i, combination + [i])

        backtrack(0, S, [])
        return result

    def remove_trips_from_combinations(self, current_max_trips_per_train):
        new_candidates = []
        copy_baseline_combinations = copy.deepcopy(self.baseline_combinations)
        for candidate in copy_baseline_combinations:
            for t in range(self.data.nb_trains):
                for i in range(len(candidate[t])):
                    if i > current_max_trips_per_train[t]-1:
                        candidate[t][i] = self.data.nb_routes
            # verify if combination is still valid
            if self.combinations.verify_daily_demands(candidate) and self.combinations.normalize_combination(candidate):
                new_candidates.append(candidate)
        return new_candidates

# ===============================================================================================================
# ===============================================================================================================

# ===============================================================================================================
# ============================= functions used to create the baseline combinations ==============================
# ===============================================================================================================

    def classify_routes(self):
        # split the cyclic routes in full cycles and short turns
        full_length = self.data.get_nb_vertices() + 1
        self.full_cycle_routes = []
        self.short_turn_routes = []
        for r in range(self.data.nb_routes):
            if not self.data.is_cyclic_route(r):
                continue
            if len(self.data.route_vertices[r]) == full_length:
                self.full_cycle_routes.append(r)
            else:
                self.short_turn_routes.append(r)

        # without full cycles, the short turns are the only cyclic bodies available
        if not self.full_cycle_routes:
            self.full_cycle_routes = list(self.short_turn_routes)

        if not self.full_cycle_routes:
            print("Instance does not have any cyclical routes...")

    def is_incomplete_route(self, r):
        # incomplete route: non-cyclic and does not pass through every point of the line
        if self.data.is_cyclic_route(r):
            return False
        points_covered = {self.data.vertex_to_point[v] for v in self.data.route_vertices[r]}
        return len(points_covered) < self.data.nb_points

    def create_pair_bodies(self):
        # two-trip short turn patterns: ordered pairs of incomplete routes where
        # each one ends at the point the other starts, closing a cycle in two trips
        pairs = []
        for a in range(self.data.nb_routes):
            if not self.is_incomplete_route(a):
                continue
            for b in range(self.data.nb_routes):
                if a == b or not self.is_incomplete_route(b):
                    continue
                if self.data.are_incompatible_routes(a, b) or self.data.are_incompatible_routes(b, a):
                    continue
                pairs.append((a, b))
        return pairs

    def create_train_sequences(self, nb_trips):
        # all valid trip sequences for a train with nb_trips trips, in two shapes:
        # an initial route followed by a repeated body, where the body is a cyclic route
        # (full cycle or short turn) or a pair of non-cyclic routes closing a short-turn
        # in two trips; and a short-turn pair opening the day, then a repeated full cycle

        initial_routes = [r for r in range(self.data.nb_routes) if self.data.is_valid_route(0, 0, r)]
        single_bodies = list(dict.fromkeys(self.full_cycle_routes + self.short_turn_routes))
        pair_bodies = self.create_pair_bodies()

        candidates = []
        for initial_route in initial_routes:
            for body in single_bodies:
                candidates.append([initial_route] + [body] * (nb_trips - 1))
            for a, b in pair_bodies:
                pattern = (a, b)
                candidates.append([initial_route] + [pattern[i % 2] for i in range(nb_trips - 1)])

        # mixed bodies -> [a, b, c, c, ...]: a short-turn pair opening the day, then a
        # repeated full cycle. Captures optima that open with a short turn and then switch
        # to full cycles (e.g. [2,3,4,4]), which a single repeated body cannot express,
        # without the blowup of enumerating arbitrary route chains
        if nb_trips >= 3:  # need the pair (2 trips) plus at least one full cycle
            for a, b in pair_bodies:
                # the pair opener must be a valid first trip (starts at the initial point)
                if not self.data.is_valid_route(0, 0, a):
                    continue
                for c in self.full_cycle_routes:
                    candidates.append([a, b] + [c] * (nb_trips - 2))

        sequences = []
        seen = set()
        for routes in candidates:
            key = tuple(routes)
            if key in seen:
                continue
            seen.add(key)

            # verify feasibility of the sequence by train
            if self.combinations.verify_sequence_feasibility(routes, self.data.nb_routes+1):
                sequences.append(routes)

        return sequences

    def create_baseline_combinations(self):
        nb_trains = self.data.nb_trains
        max_trips_per_train = self.data.max_trips_per_train
        max_nb_trips = self.data.max_nb_trips

        # if there is only 1 trip for all trains
        if max_nb_trips == 1:
            initial_routes = [r for r in range(self.data.nb_routes) if self.data.is_valid_route(0, 0, r)] # use all valid routes for the first trip (starts at the initial vertex)
            all_initial_choices = product(initial_routes, repeat=nb_trains)                               # get all combinations of initial trips for trains
            for initial_choice in all_initial_choices:
                combination = [[initial_choice[t]] for t in range(nb_trains)]

                # verify if it is a valid combination
                if (self.combinations.verify_daily_demands(combination) and self.combinations.normalize_combination(combination)):
                    self.baseline_combinations.append(combination)
            return

        self.classify_routes() # create sets of full cycles and short turns

        # create combiantions
        groups = {}
        for t in range(nb_trains):
            groups.setdefault(max_trips_per_train[t], []).append(t)                                                               # group trains according to number of trips
        group_keys = sorted(groups.keys())
        sequences_per_group = {k: self.create_train_sequences(k) for k in group_keys}                                             # create sequences of trips for each group
        all_group_choices = product(*[combinations_with_replacement(sequences_per_group[k], len(groups[k])) for k in group_keys]) # create the combinations where the order is not important 
        for group_choice in all_group_choices:
            combination = [None] * nb_trains
            for k, chosen_sequences in zip(group_keys, group_choice):
                for t, routes in zip(groups[k], chosen_sequences):
                    combination[t] = list(routes)
            # verify if it is a valid combination
            if self.combinations.verify_daily_demands(combination):
                self.baseline_combinations.append(combination)

# ===============================================================================================================
# ===============================================================================================================

# ===============================================================================================================
# ================================== functions used to solve the combinations ===================================
# ===============================================================================================================

    def solve_combination_task(self, candidate_combination):

        # verify time limit before starting to solve
        elapsed = time.time() - self.start_time
        if self.time_limit_reached or elapsed > self.time_limit_complete:
            with self.time_limit_lock:
                self.time_limit_reached = True
            with self.progress_lock:
                self.counter_solved += 1
            return

        # prune by lower bound: if the busiest train cannot beat the current best, skip the solver
        if self.combination_score(candidate_combination)[0] >= self.overall_best_sol.obj_value:
            with self.progress_lock:
                self.counter_solved += 1
                print(f"{self.counter_solved}/{len(self.candidate_combinations)} candidate combination(s) tested! - Pruned by lower bound")
            return

        if not hasattr(self.thread_local, "model"):
            self.thread_local.model = ModelTrainTimetabling(self.data, self.nb_threads, self.time_limit_complete, self.time_limit_per_combination, self.solver)
            self.thread_local.model.initialize()

        model_thread = self.thread_local.model
            
        # print(f"Solving combination: {candidate_combination}")
        model_thread.reset()
        model_thread.create_model_for_combination(candidate_combination)

        # verify time limit before executing the solver
        if self.time_limit_reached or (time.time() - self.start_time > self.time_limit_complete):
            with self.time_limit_lock:
                self.time_limit_reached = True
            with self.progress_lock:
                self.counter_solved += 1
            return
        
        feasible = model_thread.execute_solver_for_combination("heuristic", self.overall_best_sol.obj_value)

        if feasible:
            with self.best_lock:
                if model_thread.best_solution.obj_value < self.overall_best_sol.obj_value:
                    self.overall_best_sol = copy.deepcopy(model_thread.best_solution)
                    print(f"-> New best combination: {self.overall_best_sol.routes_combination} - Objective value: {self.overall_best_sol.obj_value}")
                # elif model_thread.best_solution.obj_value == self.overall_best_sol.obj_value:
                #     if model_thread.best_solution.max_nb_repeated_routes > self.overall_best_sol.max_nb_repeated_routes:
                #         self.overall_best_sol = copy.deepcopy(model_thread.best_solution)

        # verify time limit before executing the solver
        if self.time_limit_reached or (time.time() - self.start_time > self.time_limit_complete):
            with self.time_limit_lock:
                self.time_limit_reached = True
            with self.progress_lock:
                self.counter_solved += 1
            return

        with self.progress_lock:
            self.counter_solved += 1
            print(f"{self.counter_solved}/{len(self.candidate_combinations)} candidate combination(s) tested! - Feasible: {feasible}")

    def execute_candidate_combinations(self):
        self.counter_solved = 0
        futures = []
        with ThreadPoolExecutor(max_workers=self.nb_threads) as executor:
            for count in range(len(self.candidate_combinations)):
                # verify time limit before submitting new task
                if self.start_time is not None:
                    if self.time_limit_reached or (time.time() - self.start_time > self.time_limit_complete):
                        with self.time_limit_lock:
                            self.time_limit_reached = True
                        break
                
                future = executor.submit(self.solve_combination_task, self.candidate_combinations[count])
                futures.append(future)
            # if time limit was reached, try to cancel pending tasks
            if self.time_limit_reached:
                cancelled = 0
                for future in futures:
                    if future.cancel():
                        cancelled += 1
                if cancelled > 0:
                    print(f"Cancelled {cancelled} pending tasks. Waiting for running tasks to finish...")
            # wait for the tasks to finish
            executor.shutdown(wait=True)

# ===============================================================================================================
# ===============================================================================================================

# ===============================================================================================================
# ===============================================================================================================

    def run_exploration_loop(self):
        
        for S in range(1, sum(self.data.max_trips_per_train)+1, 1): # for each total ammount of trips 
            self.current_max_trips_per_train_list = self.calculate_max_trips_per_train(S, self.data.max_trips_per_train)
            for current_max_trips_per_train in self.current_max_trips_per_train_list:
                self.candidate_combinations = self.remove_trips_from_combinations(current_max_trips_per_train) # remove trips from baseline combinations
                self.candidate_combinations.sort(key=self.combination_score)                                   # solve the most promising candidates first so the cutoff tightens early
                if len(self.candidate_combinations) > 0:
                    print(f"\nTesting combinations with {current_max_trips_per_train} trips:")

                self.execute_candidate_combinations()                                                          # solve combinations

            if self.time_limit_reached:
                break

    def execute_heuristic(self):

        self.start_time = time.time()    # start counting execution time
        self.time_limit_reached = False

        self.compute_route_min_times()   # calculate minimum time of each route

        print(f"Creating baseline set of candidate combinations...")
        self.create_baseline_combinations()
        print(f"Starting exploratino loop...")
        self.run_exploration_loop()

        end_time = time.time()
        total_time = end_time - self.start_time

        # verify if time limit was reach
        self.time_limit_reached = (total_time >= self.time_limit_complete)
        if self.time_limit_reached:
            print("\n-> Time limit reached. Displaying best solution found so far...\n")
        else:
            print("\nFinished heuristic!\n")

        print(f"-> Total time = {total_time:.2f}", end="")
        self.overall_best_sol.rescale_values("heuristic")
        self.overall_best_sol.display_solution(self.data, "heuristic", self.time_limit_reached)
        self.overall_best_sol.save_solution(self.data, total_time, "heuristic", self.solver, self.time_limit_reached, self.nb_threads)
        self.overall_best_sol.create_graph(self.data, "heuristic",self.solver, self.nb_threads)

# ===============================================================================================================
# ===============================================================================================================